File: 0001-0100/0099.Recover_Binary_Search_Tree.java
//1ms

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
       TreeNode first ;
        TreeNode middle;
        TreeNode last;
        TreeNode prev;
    public void recoverTree(TreeNode root) {
        first = null;
         middle = null;
         last = null;
         prev = new TreeNode(Integer.MIN_VALUE);

        inorder(root);

        //If i got two violations that is i have first and last so swap that
        if(first!=null && last!=null){
            int temp = first.val;
            first.val = last.val;
            last.val = temp;
        }
        //If i got only one violations that is only first so swapping that with middle will correct out inorrder traversal result
        else if(first!=null && middle!=null){
            int temp = first.val;
            first.val = middle.val;
            middle.val = temp;
        }
    }
    public void inorder(TreeNode root){
        if(root == null){
            return ;
        }
        inorder(root.left);
        //Perform functons
        
        if(prev!=null && (root.val < prev.val)){
            if(first == null){
                first = prev;
                middle = root;
            }
            else{
                last = root;
            }
        } 
        prev=root;
        inorder(root.right);
    }
}

File: 0001-0100/0099.Recover_Binary_Search_Tree.cpp
//0ms


/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    TreeNode* first = nullptr;
    TreeNode* second = nullptr;
    TreeNode* prev = nullptr;

    void recoverTree(TreeNode* root) {
        inorderTraversal(root);
        swap(first->val, second->val);
    }

    void inorderTraversal(TreeNode* root) {
        if (!root) return;
        
        inorderTraversal(root->left);

        if (prev && root->val < prev->val) {
            if (!first) first = prev;
            second = root;
        }
        prev = root;

        inorderTraversal(root->right);
    }
};
